### ISSUE TITLE
[1.12][Question] Push Requests

### ISSUE THREAD

**Issue by Frontrider** [zmaster587/AdvancedRocketry#1024]
I originally wanted to make a worldgen addon, had some issues, looked into your code, than started fiddleing.

As practice, I did some modifications to the worldgen (its not quite ready, and I would also have some questions). At the moment, the moon looks a lot more believable, and exploreable.

How could I get it in front of you? Can I make a new Branch?

I would also like to do a little redesign on some parts of the code , because it can be really messy.(mostly the way you register stuff)

**Comment by zmaster587** [comment]
If you want to integrate it into AR, a new branch would be by far the easiest way to do it.  Redesigning the registration code may be a bit tricky, there are a lot of sublties and workarounds that are employed to make things work smoothly.  Most of it being client server weirdness and things needing to be registered at certain times in order to not break compatability with other mods and such.

**Comment by Frontrider** [comment]
Well, in the registration code I mostly meant creating a separate class and moving them there.
Than call the method at the point where it needs to be registered, so theres no need to scroll through all the blocks to find the tile entities. That is what I meant. Its really bad to read, and annoying to work with. The order of calls remains the same.

Also, I ported all of the old achivements which can work on the vanilla advancement system.

I'll probably mostly work on the worldgen part (it seriously needs some love), and to keep it clean I'll probably just do it on a separate branch.

Its pretty basic at the moment tough, since one reason why I started to work in AR is because I wanted to see more in action, than I realised that it should just be in ar by default.
I had a mars like (cold not hot!) and a moon type sorta ready.

Are the planet textures final (the round ones)? They look good and everything, but they still look very odd next to a square sun/moon. 😕 (a similar style would work the best)

**Comment by Frontrider** [comment]
I can not create a new branch.
Push failed: Failed with error: unable to access 'https://zmaster587@github.com/zmaster587/AdvancedRocketry.git/': The requested URL returned error: 403

Seems to be access releated, can I get a branch, where I can push?

**Comment by zmaster587** [comment]
Oh, ok.  I'd been meaning to do that for a while but then got slammed with bug reports which take higher priority than feature additions/cosmetic/ code readablity changes (unless it directly relates to the issue).

the funny thing about marslike is that is was originally going to be mars but the biome gen worked much better for Venus-like than Mars so the naming became inconsistant.  The worry with making it cold is that it may cause snow to form.

You'd need to fork the repo first, then you'd be able to create branches, from there you can create a PR.

The round textures are final, yes.  I didn't want to override the vanilla sun, but square planets also look kinda funky.  They were originally square and that didn't feel like it worked well.

**Comment by Frontrider** [comment]
I see. I'll make a resource pack then. Not a real priority.Guess it also depends on the way its approached.

Thanks for pointing me to the right direction.

I see snow forming already on my lunar hills, I'll try to figure something for that. Turning off weather works as for one type of snow.
Mostly what it would mean is spawning ice and snow during worldgen.

**Comment by zmaster587** [comment]
I think there's a flag in the biome that prevents snow from forming, I know I had to fix that for space stations and the moon, I can see about that for the hot-dry biome too.

**Comment by Frontrider** [comment]
You disabled downfall (I know about that), but after increasing the height of the terrain snow formed on top.

**Comment by zmaster587** [comment]
I think it's the canSnowAt check in the provider.  More specifically the lack of the overrride in AR.  the default is this:
```    public boolean canSnowAtBody(BlockPos pos, boolean checkLight)
    {
        Biome biome = this.getBiome(pos);
        float f = biome.getFloatTemperature(pos);

        if (f >= 0.15F)
        {
            return false;
        }
        else if (!checkLight)
        {
            return true;
        }
        else
        {
            if (pos.getY() >= 0 && pos.getY() < 256 && this.getLightFor(EnumSkyBlock.BLOCK, pos) < 10)
            {
                IBlockState iblockstate1 = this.getBlockState(pos);

                if (iblockstate1.getBlock().isAir(iblockstate1, this, pos) && Blocks.SNOW_LAYER.canPlaceBlockAt(this, pos))
                {
                    return true;
                }
            }

            return false;
        }
    }```

**Comment by Frontrider** [comment]
Snow isn't really avoided. It was still showing up on higher mountains. (the height can be beliveable, since the gravity is lower) I'll check if its still a thing.

An idea what I've been toying with, is instead of a pure tempature based system, we could have a "planet definition" system.
-Every planet type has its tempature ranges where it can spawn in.
-It can be configured that it would only spawn as a moon or as a planet.(or booth)
-Size range could be configured. For gravity, and also to determine height variations, since lower gravity means higher mountains can exist. (In the auto biome select, we can just blacklist jungles if its above a certain gravity)
-Pre-selected biomes ,or you can set it to autoselect, like now. Even to just use a subset of biomes from a list. So even same-ish planets get a chance to become different
-You no longer blacklist biomes, instead you put them on the whitelist of specific planet types. (can be configured)

I'm also trying a design for a way to fake custom biomes, whitout ever hitting the biome id limit, but its very very early. Before this:

I couldn't see a "default" biome class in your branch, if its still not a thing, my branch has a "BiomeBase" class inside the biomes folder, which has some custom settings to it, and I will add more later. (its a little buggy, but it works.

I have some refactoring, what I you will get in a pr this week, when I manage to chew through all of it. Its mostly stuff like moving registrations outside the main class because a lot of it, really doesn't need to be there, so it becomes easier to read.
You really went overboard with that, and it makes it harder to work with.(now Block registration is inside the blocks class, Item registration is inside the items class etc...) So, there won't be a need to scroll through every block, item and tileentity when I want to register a Biome. That stuff has nothing to do with any Mod compat based ordering, they are called by forge they work just fine outside the main class.

I saw the modid as a defined string inside every line of bock registration, now I reference the modid from the main class, and it actually feels like it loads faster. Java Strings can be heavy.

**Comment by zmaster587** [comment]
-The current system is not too different, it basically works in reverse of that.  Rather than picking a type then calculating everything else, it picks a few random variables then calculates the rest and determines the type from that.  Temperature is derived from orbital distance from the host star, the host star's temperature, and the planet's atmosphere density.  There's also already a check for high pressure for special biome handling, (see highpressurebiome list, stormlands, deepswamp).  So I think it's pretty well covered as is in that category.  Though there does need to be low pressure.  I suppose fuzzy logic for biome spawning could be an enhancement though.
- I don't quite follow spawning as a moon vs planet.  Moons by definition have to orbit a planet.
- Gravity determining height variations would be cool to see, could be easily implemented as a multiplier on the already existing heightscale field in ChunkProviderPlanet.generateHeightmap().  It may also be cool to play around with the X and Z scale factors as well.  May give rise to large rolling hills or many steep crevices.

- I believe that the randomization be already exists here https://github.com/zmaster587/AdvancedRocketry/blob/1.12/src/main/java/zmaster587/advancedRocketry/dimension/DimensionProperties.java#L924 (and it turns out the config to change the number of biomes per world doesn't actually do anything.. oops).  Though having user defined biomes per category, perhaps similar to the ore gen config may be nice to have.

-  The problem I have with not being able to black list biomes is that there's a very common usecase:  BOP (or adding a mod that adds one undesirable biome like AE).  Many pack makers probably want all viable biomes to spawn except for a select few, and forcing them to have to manually define all of them just to remove one or two biomes is a PITA.  The other issue is using a white list removes the "it just works" aspect with dropping new biome adding mods in.  Though i would not be opposed to the idea of an optional whitelist, if it exists or contains at least one entry, then it overrides default and blacklist.  Doing that would allow fine configuration for those that want it, but not making it a PITA for everyone else.

I'd toyed with the idea in the past actually.  One of the methods I looked into was using a secondary 2D short array that contained extra biome info and was stored in the chunk.  This allowed up to 65526 + 256 biomes. one of the problems I ran nto though was a significant increase in RAM usage and filesize.  The other option ofc is to generate it on the fly, but that doesn't fix the RAM usage issue.

Another alternative I investigated was simply not storing the biome since the properties were simply used for world gen really and assigning a similar, existing vanilla biome for the sky color, rain, humidity and temp properties.

I'm not currently using a base biome since I've not really had a need for it, they're all different and none of them really do anything that's reused, so adding a base class with the current design only added memory overhead and served no real purpose.  However depending on future developments it may be needed at some point.

I really appreciate your help with refactoring that.  I've been buried so deep in issue reports that have no relation to registration code that I've not had time to get to it myself.  Though keep in mind that any changed made do have to be backportable in structure through 1.7.10 otherwise my automated backporting tool will throw a fit every time.  For pre-1.12 a static call should suffice though, I can take care of that.

It'll use a few KB less memory perhaps, java strings are heavy, but not heavy enough that strings less than 255 characters with less than 100 occurrences will make a noticeable change unless something else is deeply broken.  Even systems dating before 2010 can chew through 500,000+ characters worth of concatenation, comparing and copying in less than a second XD
